Add tooltip to CopyToClipboard

rebeccaalpert opened the pull request:
Fixes [#144](https://github.com/openshift/console/issues/144).

There's now a tooltip when you mouse over the CopyToClipboard button. The regular tooltip text says "Copy to Clipboard" and the tooltip text changes to "Copied" once the button is clicked.

Not copied:
<img width="1349" alt="screen shot 2018-06-27 at 5 46 42 pm" src="https://user-images.githubusercontent.com/7014965/42001538-417e5d6a-7a32-11e8-9492-ea135910973c.png">

Copied:
<img width="1346" alt="screen shot 2018-06-27 at 5 46 47 pm" src="https://user-images.githubusercontent.com/7014965/42001545-46ff1734-7a32-11e8-9a56-c0133a03ed4a.png">


@openshift/team-ux-review, could you please review this?

coreos-ui commented:
Can one of the admins verify this patch?

coreos-ui commented:
Can one of the admins verify this patch?

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+  }
+
+  render() {
+    const overrides = {
```
If this can be reused in other contexts for a left-facing tooltip, I'd move it to tooltip.jsx as another option (top by default, but left position either as a prop or another named component)

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+        <pre className="co-pre-wrap co-copy-to-clipboard__text">{this.props.visibleValue}</pre>
+        <Tooltip content={text} styles={overrides}>
+          <CTC text={this.props.value} onCopy={() => this.setState({copied: true})}>
+            <button onMouseEnter={() => this.setState({copied: false})} className="btn btn-default co-copy-to-clipboard__btn fix" type="button">
```
I'd make this a named function in the class and bind to this in the contructor

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+      text = <span className="co-nowrap" key="nowrap">Copy to Clipboard</span>;
+    }
+
+    return <React.Fragment>
```
Not new, but this shouldn't need React.Fragment since you're only returning one div.

spadgett commented:
@rhamilto can you review the styles?

spadgett commented:
jenkins rebuild

rhamilto commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+    const overrides = {
+      wrapper: {
+        position: 'absolute',
+        right: '0px',
```
Since zero is always zero regardless of unit, you can drop "px" from the value here and on line 17.

rhamilto commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+      },
+      arrow: {
+        bottom: '8px',
+        left: '103%',
```
```
left: 'auto',
leftMargin: 'auto',
right: '-13px',
```
instead?  That's a bit more precise (8px arrow + 5px tooltip padding = 13px).

rhamilto commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+      },
+      content: {
+        display: 'block',
+        fontSize: '12px',
```
add `fontFamily: '"Open Sans",Helvetica,Arial,sans-serif',` as the current font is the monospace stack that's being inherited from `<pre>`.

spadgett replied:
If the styles really are specific to this component, I'd be OK keeping them in this file just for the copy button instead of trying to generalize them.

rhamilto commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,65 @@
+    return <React.Fragment>
+      <div className="co-copy-to-clipboard">
+        <pre className="co-pre-wrap co-copy-to-clipboard__text">{this.props.visibleValue}</pre>
+        <Tooltip content={text} styles={overrides}>
```
Need to pass `text` as an array since it contains markup (e.g., `content={[text]}`).  Note the console error.

![screen shot 2018-06-28 at 9 34 50 am](https://user-images.githubusercontent.com/895728/42037513-906e0e46-7ab6-11e8-955f-bfd13020433f.PNG)

rebeccaalpert commented:
I reworked the styles so that the tooltips can accommodate text of different lengths. I also updated the functions, text prop, etc.

Not copied:
![screen shot 2018-06-28 at 2 49 59 pm](https://user-images.githubusercontent.com/7014965/42054866-8a85ec7e-7ae3-11e8-8cfc-aa363635b2be.png)

Copied:
![screen shot 2018-06-28 at 2 50 02 pm](https://user-images.githubusercontent.com/7014965/42054873-8e771bb4-7ae3-11e8-92f1-ee48d4350840.png)

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,80 @@
+      text = [<span className="co-nowrap" key="nowrap">Copied</span>];
+    } else {
+      text = [<span className="co-nowrap" key="nowrap">Copy to Clipboard</span>];
+    }
```
Better to use a ternary here so we can use `const`. Something like

```jsx
const tooltipText =  this.state.copied ? 'Copied' : 'Copy to Clipboard';
const tooltipContent = [<span className="co-nowrap" key="nowrap">{text}</span>];
```

then we can pass `tooltipContent` to the tooltip.

spadgett commented on frontend/public/components/utils/copy-to-clipboard.jsx:
```diff
@@ -1,15 +1,80 @@
+  }
+
+  render() {
+    const overrides = {
```
Consider `Object.freeze` as used in tooltip.jsx

https://github.com/openshift/console/blob/master/frontend/public/components/utils/tooltip.jsx#L4

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze

rebeccaalpert commented:
I swapped in the ternary and added Object.freeze.

spadgett commented:
```
/go/src/github.com/openshift/console/frontend/public/components/utils/copy-to-clipboard.jsx
  61:26  error  Multiple spaces found before 'this'  no-multi-spaces
```

rebeccaalpert commented:
I removed the extra space and squashed my commits.

spadgett commented:
/hold

spadgett commented:
jenkins rebuild

spadgett commented:
/hold cancel
